tools/IoTEdgeDevOps/VstsPipelineSync/VstsBuildBatchUpdate.cs

```diff
@@ -0,0 +1,122 @@
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine($"Unexcepted Exception: {ex}");
+                }
+
+                Console.WriteLine($"Import Vsts Builds data finished at {DateTime.UtcNow}; wait {waitPeriodAfterEachUpdate} for next update.");
+                await Task.Delay((int)waitPeriodAfterEachUpdate.TotalMilliseconds);
+            }
+        }
+
+        private async Task<Dictionary<BuildDefinitionId, DateTime>> ImportVstsBuildsDataAsync(BuildManagement buildManagement, string branch, HashSet<BuildDefinitionId> buildDefinitionIds, Dictionary<string, Dictionary<BuildDefinitionId, DateTime>> lastUpdatePerBranchPerDefinition)
+        {
+            Dictionary<BuildDefinitionId, DateTime> lastUpdatePerDefinition = lastUpdatePerBranchPerDefinition.GetIfExists(branch);
+            if (lastUpdatePerDefinition == null)
+            {
+                lastUpdatePerDefinition = new Dictionary<BuildDefinitionId, DateTime>();
+            }
+
+            foreach (BuildDefinitionId buildDefinitionId in buildDefinitionIds)
+            {
+                DateTime maxLastChange = await ImportVstsBuildsDataForSpecificDefinitionAsync(buildManagement, branch, buildDefinitionId, lastUpdatePerDefinition.GetIfExists(buildDefinitionId));
+                lastUpdatePerDefinition.Upsert(buildDefinitionId, maxLastChange);
+            }
+
+            return lastUpdatePerDefinition;
+        }
+
+        private async Task<DateTime> ImportVstsBuildsDataForSpecificDefinitionAsync(
+            BuildManagement buildManagement,
+            string branch,
+            BuildDefinitionId buildDefinitionId,
+            DateTime lastUpdate)
+        {
+            SqlConnection sqlConnection = null;
+
+            try
+            {
+                sqlConnection = new SqlConnection(this.dbConnectionString);
+                sqlConnection.Open();
+
+                IList<VstsBuild> buildResults = await buildManagement.GetBuildsAsync(new HashSet<BuildDefinitionId> { buildDefinitionId }, branch, lastUpdate);
+                Console.WriteLine($"Query VSTS for branch [{branch}] and build definition [{buildDefinitionId.ToString()}]: last update={lastUpdate} => result count={buildResults.Count}");
+                DateTime maxLastChange = DateTime.MinValue;
+
+                foreach (VstsBuild build in buildResults.Where(r => r.HasResult()))
+                {
+                    var cmd = new SqlCommand
+                    {
+                        Connection = sqlConnection,
+                        CommandType = System.Data.CommandType.StoredProcedure,
+                        CommandText = "UpsertVstsBuildRun"
+                    };
+
+                    cmd.Parameters.Add(new SqlParameter("@BuildNumber", build.BuildNumber));
+                    cmd.Parameters.Add(new SqlParameter("@DefinitionId", build.DefinitionId));
```

Comment:
We are going to need the readable name here

Reply:
done